Metadata Table recovery failure

**Describe the bug**
Ran cingest on a 2.1.0-SNAPSHOT cluster with per-table crypto configured. Specifically, the following properties are set in `accumulo.properties`:
```
instance.crypto.opts.factory=org.apache.accumulo.core.spi.crypto.PerTableCryptoServiceFactory
general.custom.crypto.recovery.service=org.apache.accumulo.core.spi.crypto.AESCryptoService
general.custom.crypto.wal.service=org.apache.accumulo.core.spi.crypto.AESCryptoService
general.custom.crypto.key.uri=file:///opt/accumulo-testing/accumulo/accumulo-2.1.0-SNAPSHOT/conf/per-table-enc-key
```
and the `ci` table had the property `table.crypto.opts.service=org.apache.accumulo.core.spi.crypto.AESCryptoService`. I loaded the table with a bunch of data, then I deleted it via the shell. The `deletetable` command never returned and it appears that one of the metadata tablets was re-assigned during the delete operation. When trying to load the metadata tablet, the tserver logged the following error:

```
2022-09-28T16:50:44,468 [log.RecoveryLogsIterator] DEBUG: Opening recovery log dir 11946b4a-99b6-4b7e-b0d4-ccc1f981783a
2022-09-28T16:50:44,469 [crypto.CryptoFactoryLoader] DEBUG: Creating new crypto factory class org.apache.accumulo.core.spi.crypto.GenericCryptoServiceFactory
2022-09-28T16:50:44,472 [tserver.AssignmentHandler] WARN : exception trying to assign tablet !0<;~ null
java.lang.RuntimeException: Error recovering tablet !0<;~ from log files
        at org.apache.accumulo.tserver.tablet.Tablet.<init>(Tablet.java:379) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.AssignmentHandler.run(AssignmentHandler.java:165) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.ActiveAssignmentRunnable.run(ActiveAssignmentRunnable.java:63) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.trace.TraceWrappedRunnable.run(TraceWrappedRunnable.java:52) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at org.apache.accumulo.core.trace.TraceWrappedRunnable.run(TraceWrappedRunnable.java:52) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: java.io.IOException: java.io.UncheckedIOException: java.io.IOException: incorrect header check
        at org.apache.accumulo.tserver.log.TabletServerLogger.recover(TabletServerLogger.java:544) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.TabletServer.recover(TabletServer.java:1204) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.tablet.Tablet.<init>(Tablet.java:334) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        ... 7 more
Caused by: java.io.UncheckedIOException: java.io.IOException: incorrect header check
        at org.apache.accumulo.core.file.blockfile.impl.CachableBlockFile$Reader$BCFileLoader.load(CachableBlockFile.java:223) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.client.rfile.RFileScanner$NoopCache.getBlock(RFileScanner.java:165) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.client.rfile.RFileScanner$NoopCache.getBlock(RFileScanner.java:158) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.file.blockfile.impl.CachableBlockFile$Reader.getMetaBlock(CachableBlockFile.java:376) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.file.rfile.RFile$Reader.<init>(RFile.java:1164) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.file.rfile.RFile$Reader.<init>(RFile.java:1256) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.core.client.rfile.RFileScanner.iterator(RFileScanner.java:354) ~[accumulo-core-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.log.RecoveryLogsIterator.validateFirstKey(RecoveryLogsIterator.java:156) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.log.RecoveryLogsIterator.<init>(RecoveryLogsIterator.java:77) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.log.SortedLogRecovery.findMaxTabletId(SortedLogRecovery.java:107) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.log.SortedLogRecovery.findLogsThatDefineTablet(SortedLogRecovery.java:147) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.log.SortedLogRecovery.recover(SortedLogRecovery.java:291) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.log.TabletServerLogger.recover(TabletServerLogger.java:542) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.TabletServer.recover(TabletServer.java:1204) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        at org.apache.accumulo.tserver.tablet.Tablet.<init>(Tablet.java:334) ~[accumulo-tserver-2.1.0-SNAPSHOT.jar:2.1.0-SNAPSHOT]
        ... 7 more
Caused by: java.io.IOException: incorrect header check
        at org.apache.hadoop.io.compress.zlib.ZlibDecompressor.inflateBytesDirect(Native Method) ~[hadoop-client-api-3.3.4.jar:?]
        at org.apache.hadoop.io.compress.zlib.ZlibDecompressor.decompress(ZlibDecompressor.java:225) ~[hadoop-client-api-3.3.4.jar:?]
...
```

One tablet for the metadata table is in an offline state and the `ci` table is in a deleting state.

--- a/server/tserver/src/main/java/org/apache/accumulo/tserver/log/RecoveryLogsIterator.java
+++ b/server/tserver/src/main/java/org/apache/accumulo/tserver/log/RecoveryLogsIterator.java
@@ -21,16 +21,21 @@ package org.apache.accumulo.tserver.log;
 import java.io.IOException;
 import java.util.AbstractMap;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map.Entry;
 
-import org.apache.accumulo.core.client.Scanner;
-import org.apache.accumulo.core.client.ScannerBase;
 import org.apache.accumulo.core.client.rfile.RFile;
+import org.apache.accumulo.core.crypto.CryptoEnvironmentImpl;
 import org.apache.accumulo.core.data.Key;
 import org.apache.accumulo.core.data.Range;
 import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.file.FileOperations;
+import org.apache.accumulo.core.file.FileSKVIterator;
+import org.apache.accumulo.core.iterators.IteratorAdapter;
+import org.apache.accumulo.core.spi.crypto.CryptoEnvironment;
+import org.apache.accumulo.core.spi.crypto.CryptoService;
 import org.apache.accumulo.server.ServerContext;
 import org.apache.accumulo.server.fs.VolumeManager;
 import org.apache.accumulo.server.log.SortedLogState;
@@ -53,7 +58,7 @@ public class RecoveryLogsIterator
 
   private static final Logger LOG = LoggerFactory.getLogger(RecoveryLogsIterator.class);
 
-  private final List<Scanner> scanners;
+  private final List<FileSKVIterator> fileIters;
   private final Iterator<Entry<Key,Value>> iter;
 
   /**
@@ -63,7 +68,7 @@ public class RecoveryLogsIterator
       LogFileKey end, boolean checkFirstKey) throws IOException {
 
     List<Iterator<Entry<Key,Value>>> iterators = new ArrayList<>(recoveryLogDirs.size());
-    scanners = new ArrayList<>();
+    fileIters = new ArrayList<>();
     Range range = start == null ? null : LogFileKey.toRange(start, end);
     var vm = context.getVolumeManager();
 
@@ -77,20 +82,26 @@ public class RecoveryLogsIterator
         validateFirstKey(context, fs, logFiles, logDir);
       }
 
-      for (Path log : logFiles) {
-        var scanner = RFile.newScanner().from(log.toString()).withFileSystem(fs)
-            .withTableProperties(context.getConfiguration()).build();
+      CryptoEnvironment env = new CryptoEnvironmentImpl(CryptoEnvironment.Scope.RECOVERY);
+      CryptoService cryptoService = context.getCryptoFactory().getService(env,
+          context.getConfiguration().getAllCryptoProperties());
 
-        scanner.setRange(range);
-        Iterator<Entry<Key,Value>> scanIter = scanner.iterator();
+      for (Path log : logFiles) {
+        FileSKVIterator fileIter = FileOperations.getInstance().newReaderBuilder()
+            .forFile(log.toString(), fs, fs.getConf(), cryptoService)
+            .withTableConfiguration(context.getConfiguration()).seekToBeginning().build();
+        if (range != null) {
+          fileIter.seek(range, Collections.emptySet(), false);
+        }
+        Iterator<Entry<Key,Value>> scanIter = new IteratorAdapter(fileIter);
 
         if (scanIter.hasNext()) {
           LOG.debug("Write ahead log {} has data in range {} {}", log.getName(), start, end);
           iterators.add(scanIter);
-          scanners.add(scanner);
+          fileIters.add(fileIter);
         } else {
           LOG.debug("Write ahead log {} has no data in range {} {}", log.getName(), start, end);
-          scanner.close();
+          fileIter.close();
         }
       }
     }
@@ -115,8 +126,10 @@ public class RecoveryLogsIterator
   }
 
   @Override
-  public void close() {
-    scanners.forEach(ScannerBase::close);
+  public void close() throws IOException {
+    for (FileSKVIterator fskv : fileIters) {
+      fskv.close();
+    }
   }
 
   /**

--- a/server/tserver/src/main/java/org/apache/accumulo/tserver/log/RecoveryLogsIterator.java
+++ b/server/tserver/src/main/java/org/apache/accumulo/tserver/log/RecoveryLogsIterator.java
@@ -26,7 +26,6 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.Map.Entry;
 
-import org.apache.accumulo.core.client.rfile.RFile;
 import org.apache.accumulo.core.crypto.CryptoEnvironmentImpl;
 import org.apache.accumulo.core.data.Key;
 import org.apache.accumulo.core.data.Range;
@@ -60,6 +59,7 @@ public class RecoveryLogsIterator
 
   private final List<FileSKVIterator> fileIters;
   private final Iterator<Entry<Key,Value>> iter;
+  private final CryptoEnvironment env = new CryptoEnvironmentImpl(CryptoEnvironment.Scope.RECOVERY);
 
   /**
    * Scans the files in each recoveryLogDir over the range [start,end].
@@ -72,19 +72,18 @@ public class RecoveryLogsIterator
     Range range = start == null ? null : LogFileKey.toRange(start, end);
     var vm = context.getVolumeManager();
 
+    final CryptoService cryptoService = context.getCryptoFactory().getService(env,
+        context.getConfiguration().getAllCryptoProperties());
+
     for (Path logDir : recoveryLogDirs) {
       LOG.debug("Opening recovery log dir {}", logDir.getName());
       List<Path> logFiles = getFiles(vm, logDir);
       var fs = vm.getFileSystemByPath(logDir);
 
       // only check the first key once to prevent extra iterator creation and seeking
-      if (checkFirstKey) {
-        validateFirstKey(context, fs, logFiles, logDir);
+      if (checkFirstKey && !logFiles.isEmpty()) {
+        validateFirstKey(context, cryptoService, fs, logFiles, logDir);
       }
-
-      CryptoEnvironment env = new CryptoEnvironmentImpl(CryptoEnvironment.Scope.RECOVERY);
-      CryptoService cryptoService = context.getCryptoFactory().getService(env,
-          context.getConfiguration().getAllCryptoProperties());
 
       for (Path log : logFiles) {
         FileSKVIterator fileIter = FileOperations.getInstance().newReaderBuilder()
@@ -161,12 +160,13 @@ public class RecoveryLogsIterator
   /**
    * Check that the first entry in the WAL is OPEN. Only need to do this once.
    */
-  private void validateFirstKey(ServerContext context, FileSystem fs, List<Path> logFiles,
-      Path fullLogPath) {
-    try (var scanner =
-        RFile.newScanner().from(logFiles.stream().map(Path::toString).toArray(String[]::new))
-            .withFileSystem(fs).withTableProperties(context.getConfiguration()).build()) {
-      Iterator<Entry<Key,Value>> iterator = scanner.iterator();
+  private void validateFirstKey(ServerContext context, CryptoService cs, FileSystem fs,
+      List<Path> logFiles, Path fullLogPath) throws IOException {
+    try (FileSKVIterator fileIter = FileOperations.getInstance().newReaderBuilder()
+        .forFile(logFiles.get(0).toString(), fs, fs.getConf(), cs)
+        .withTableConfiguration(context.getConfiguration()).seekToBeginning().build()) {
+      Iterator<Entry<Key,Value>> iterator = new IteratorAdapter(fileIter);
+
       if (iterator.hasNext()) {
         Key firstKey = iterator.next().getKey();
         LogFileKey key = LogFileKey.fromKey(firstKey);

